README.md
# UFTP - Encrypted UDP based FTP with multicast
(Cloned from https://uftp-multicast.sourceforge.net)

UFTP is an encrypted multicast file transfer program, designed to securely,
reliably, and efficiently transfer files to multiple receivers simultaneously.
This is useful for distributing large files to a large number of receivers,
and is especially useful for data distribution over a satellite link (with two
way communication), where the inherent delay makes any TCP based communication
highly inefficient.  The multicast encryption scheme is based on TLS with
extensions to allow multiple receivers to share a common key.

UFTP also has the capability to communicate over disjoint networks separated
by one or more firewalls (NAT traversal) and without full end-to-end multicast
capability (multicast tunneling) through the use of a UFTP proxy server. 
These proxies also provide scalability by aggregating responses from a group
of receivers.

## Building

UNIX-like systems require GNU make and a C compiler such as GCC or
equivalent.  Windows systems require Visual Studio Community 2015 or later.

Non-Windows systems require OpenSSL to be installed if encryption support is
enabled.  On Linux, Solaris, and BSD systems (including MacOSX), this should
be included with the OS.


To compile for UNIX-like systems, including MacOSX:
`make [ OPENSSL={install directory for OpenSSL} ] [ NO_ENCRYPTION=1 ]`

To compile for Windows (from a Visual Studio command prompt):
`nmake -f makefile.mak [ OPENSSL={install directory for OpenSSL} ] [ NO_ENCRYPTION=1 ]`

By default, Visual Studio compiles a 32-bit executable.  To compile in 64-bit
mode, first cd to the VC subdirectory under the Visual Studio install directory.
Then run the following command:

`vcvarsall amd64`

The OPENSSL parameter to make should only need to be specified if OpenSSL is
installed in a non-standard location, or on systems where it isn't
preinstalled.

The NO_ENCRYPTION flag compiles with no encryption support.  This can be
useful in embedded environments that don't need encryption and want to
keep the size of the executable down, and for use in a Windows service
that doesn't require encryption.


To install for UNIX-like systems, including MacOSX:
`make [ DESTDIR={install directory} ] install`

The DESTDIR parameter allows installing into a fake root directory, which
can be useful for packaging utilities such as rpm.


## Tuning

If you find that clients can't receive data fast enough, or if servers
communicating with several hundred clients can't handle the flood of
STATUS messages that come in, you can increase the UDP send/receive buffer
size to help with this.  This is set using the -B option on server, 
client, or proxy.  However, many operating systems impose a maximum value
of 256K (262144).  This limit can be increased, although the method is OS
specific.  Here are a few common ones:

Solaris: `ndd -set /dev/udp udp_max_buf {value}`

Linux: `sysctl -w net.core.rmem_max={value}`
       `sysctl -w net.core.wmem_max={value}`

MacOSX / FreeBSD: `sysctl -w kern.ipc.maxsockbuf={value}`
(The actual maximum you can use is 8/9 of this value)

While Windows does not have this limitation, it does throttle back UDP
datagrams larger that 1024 bytes by default (a UFTP packet is 1472 bytes by
default).  This is most commonly seen when attempting to move data at 100
Mbps or more.  You can change this by adding/modifying the following registry
value:

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\AFD\Parameters\FastSendDatagramThreshold

Set this DWORD value to 1500 and reboot, and it should take care of the issue.

If you're making use of the DSCP/TOS byte in the IP header, this can be
modified via the -Q option to the server, client, or proxy.  Setting this byte
is not currently supported in Windows.

The TOS byte also contains the ECN bits, which are used by the TFMCC 
congestion control scheme, so restrictions on the TOS byte also affect the
sender support for ECN.

## Upgrading from 4.x to 5.x

Key changes from 4.x to 5.x:

- The use of source specific multicast is not automatically turned on when
  a client or proxy uses a server list file.  It is now enabled separately
  with the -o option on the client and the proxy.
- Under 4.x, a client proxy would substitute its own keys in place of the
  server's when forwarding an ANNOUNCE.  Now the ANNOUNCE is forwarded
  unchanged and the proxy sends a PROXY_KEY with its own keys.
- Because of the key changes, clients are now fully aware of proxies.  If a
  client talks to a server through a proxy, the server must be listed in the
  server list file along with the ID of proxy it communicates through.
- The server list file no longer contains the key fingerprint of the proxy
  that the server goes through but the fingerprint of the server itself.
  Fingerprints for proxies are now specified in a separate file.

The client and proxy support both the version 4 and version 5 protocols.
When upgrading from version 4.x to version 5.x, the following steps should
be taken:

If you use encrypted sessions and are running version 4.10.1 or earlier,
the server should first be upgraded to version 4.10.2.  Also, since 5.x
versions have a more restricted set of encryption parameters, the server's
parameters should be updated to match.  Specifically, it must use an ECDH key
exchange mode, a GCM or CCM mode symmetric cypher, and SHA-256, SHA-384,
or SHA-512 for the hash.

If proxies are not in use, upgrade clients one at a time to 5.x, then upgrade
the server to 5.x once all clients are upgraded.

If proxies are in use and you want to perform a rolling upgrade:

- Upgrade server proxies to 5.x one at a time.
- Upgrade client proxies to 5.x one at a time.
- Upgrade response proxies to 5.x one at a time.
- Upgrade clients to 5.x one at a time.  If the client is using client 
  proxies, they must be listed in the server list file specified by -S for
  each server that uses a proxy.  Also, if your server list file contains key
  fingerprints, you will need to add an additional entry for each server.
  Under 4.x a server entry contains the fingerprint of the proxy used by the
  server, while under 5.x the entry contains the server's fingerprint
  with the proxy's fingerprint in a separate file.  These servers will need
  to be assigned a new ID when upgrading so that clients can have two server
  list entries for the same server: one with the current ID running 4.x and the
  proxy's fingerprint, and one with the new ID running 5.x with the server's
  fingerprint.
- Upgrade server to 5.x.


Changes.txt
Version 5.0.1 - 8/2/2022
  On very low speed transfers (<10Kbps) sessions would time out due to a very
    large interpacket transmission interval.  Fixed by putting a lower limit
    on the advertised GRTT of of the interpacket transmission interval.
  Sending of ABORT messages on early shutdown would sometimes fail due to
    OpenSSL cleanup functions running before application cleanup.  Changed the
    ordering of atexit() handlers to ensure OpenSSL cleanup happens last.
  Fixed missing timestamp update when clients read CONG_CTRL messages
  Fix to GRTT handling on server to ensure it doesn't fall below minumim.
  Fixed bypassed checking of existing files on client for backup
  Various logging fixes

Version 5.0 - 4/22/2020
  Major security updates.  The key exchange and key derivation algorithms
    were modified and supported algorithms were pruned using TLS 1.3 as a
    basis.  This includes:
  - HKDF used in multiple stages for key derivation from raw shared secrets.
  - Included addtional context in key derivation and signatures to protect
    against replay attacks and downgrade attacks.
  - Reduced set of supported EC curves to those supported by TLS 1.3
  - Removed RSA key exchange which does not provide perfect forward secrecy.
    All key exchanges now use ECDH.
  - Removed support for SHA-1 hashes in key exchanges.
  - Supported symmetric ciphers are AES in AEAD mode (GCM or CCM).
  - Increased supported RSA key sizes
  Encrypted sessions are now enabled by default.  It can be disabled by
    specifying "none" for the key type in the server's -Y option.
  Backward compatibility retained for version 4.x in clients and proxies.
    When communicating with a 4.x server, only allow algorithms and key
    exchange modes permitted in the new version.
  Clients and proxies no longer need to use signature keys that match the
    type and size used by the server.  As a result, the -k and -K options to
    the client now only accept a single key instead of multiple.  The proxy
    still supports multiple keys for 4.x compatibility, however only the first
    key listed is used for any version 5.x session.
  Proxies now send their keys in a separate message instead of injecting them
    in the ANNOUNCE sent by the server.  This allows clients to be fully
    aware of proixes and allows them to authenticate servers and proxies
    separately.
  Format of client's server list modified to specify the proxy that a server
    communicates through.  Fingerprints listed in this file now always
    specify the server as opposed to having the proxy's key in some cases.
  Added -R option to client to specify a list of proxies along with their
    public key fingerprints.  The old use of -R to specify a version 4.x
    response proxy has moved to -r.
  Previously, using -S in the client or proxy to specify a server list would
    automatically enable source specific multicast (SSM).  The use of SSM is
    now enabled separately via the -o option on both the client and proxy.
  Fixed a bug that caused ECDSA signatures created on Linux with curve
    secp521r1 from being verified successfully on Windows.
  Fixed cleanup on clients and proxies to prevent occasional crashes on
    shutdown under Windows.
  Update timstamps in messages to use 64-bit microseconds since the epoch,
    addressing Y2038 issues.

Version 4.10.2 - 4/5/2020
  Fixed security issue where using sha384 or sha512 would set encryption keys
    to all bytes 0
  When using ECDH key exchange with closed group membership, an incorrect
    signature would be applied to the ANNOUCE message, causing the session
    to fail.  Bug fixes.
  Relaxed server side checks on the type of key supplied by a client when not
    using public key signatures on all messages.  This will assist in the
    upgrade process to the upcoming version 5.0.
  Fixed various small memory leaks

Version 4.10.1 - 1/15/2020
  Previously the timestamp of received files would only be set in sync mode.
    This has been fixed to always set the timestamp.
  Status file result messages were not being printed in all cases.  Bug fixed.
  Reject non-encrypted CONG_CTRL and CC_ACK messages when using encryption.

Version 4.10 - 7/10/2019
  Allow specifying transmission rates of up to 1 Tbps.  This involved changing
    a number of variables tracking the rate from 32 bit to 64 bit types.  It
    also changed the tracking of wait times between data packet transmissions
    from microseconds to nanoseconds.
  Fixed a bug when getting the list of network interfaces in the event that
    an interface does not have an interface index number.

Version 4.9.11 - 5/13/2019
  The error check added in 4.9.10 on chdir wasn't checking the correct value,
    causing the client and proxy to not start in background mode.  Bug fixed.

Version 4.9.10 - 5/9/2019
  The initial CLR selection was picking the receiver with the lowest RTT
    instead of the higest.  Fixed to select the receiver with the highest RTT.
  While the server is sending, check for CTRL-C by the user at least once a
    second instead of waiting for a GRTT cycle if it is longer.
  Fixed rate calculation bug when TFMCC is specified with -R -1.
  Minor error checking / log format fixes

Version 4.9.9 - 2/17/2019
  Fixed handling of min_time in TFMCC mode

Version 4.9.8 - 8/29/2018
  When compiling with MSVC 2017, the IP_MTU_DISCOVER flag is defined but not
    implemented, resulting in errors when attempting to set it.  Added check 
    for the case if the flag is not implemented.
  Machines with a blank IP address for the first interface were getting a
    default UID of 0.  When this happens, check for a non-blank interface
    to derive the UID from.
  Cleaned up setting of key exchange random numbers to avoid aliasing violation

Version 4.9.7 - 5/27/2018
  Fixed delay in server sending thread causing transfer of large files to
    time out.  The block number to send was being incremented once per

uftp.h
/*
 *  UFTP - UDP based FTP with multicast
 *
 *  Copyright (C) 2001-2022   Dennis A. Bush, Jr.   bush@tcnj.edu
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  Additional permission under GNU GPL version 3 section 7
 *
 *  If you modify this program, or any covered work, by linking or
 *  combining it with the OpenSSL project's OpenSSL library (or a
 *  modified version of that library), containing parts covered by the
 *  terms of the OpenSSL or SSLeay licenses, the copyright holder
 *  grants you additional permission to convey the resulting work.
 *  Corresponding Source for a non-source form of such a combination
 *  shall include the source code for the parts of OpenSSL used as well
 *  as that of the covered work.
 */

#ifndef _UFTP_H
#define _UFTP_H

#include <stdint.h>

#ifdef WINDOWS

// same as passing /D _CRT_SECURE_NO_WARNINGS to cl
#pragma warning(disable: 4996)

#include <windows.h>
#include <process.h>

#define open(name, ...) _open(name, __VA_ARGS__)
#define read(fd, buf, count) _read(fd, buf, count)
#define close(fd) _close(fd)
#define write(fd, buf, count) _write(fd, buf, count)
#define dup2(fd1, fd2) _dup2(fd1, fd2)
#define unlink(file) _unlink(file)
#define rmdir(dir) _rmdir(dir)
#define getpid() _getpid()
#define mkdir(dir, mode) _mkdir(dir)
#define usleep(t) Sleep((int)(t)/1000)
#define sleep(t) Sleep((int)(t)*1000)
#define strdup(p) _strdup(p)
#define utime(f, t) _utime64(f, t)
#define isfullpath(str) (((str[1] == ':') && (str[2] == '\\')) || \
                         ((str[0] == '\\') && (str[1] == '\\')))
#define PATH_SEP '\\'

#define F_i64 "%I64d"
#define F_u64 "%I64u"

typedef uint64_t f_offset_t;
typedef struct _stat64 stat_struct;
typedef struct __utimbuf64 utim_buf;
#define stat_func(name, buf) _stat64(name, buf)
#define lstat_func(name, buf) _stat64(name, buf)
#define S_ISCHR(mode)   (((mode) & S_IFMT) == S_IFCHR)
#define S_ISDIR(mode)   (((mode) & S_IFMT) == S_IFDIR)
#define S_ISREG(mode)   (((mode) & S_IFMT) == S_IFREG)
#define lseek_func(fd, offset, whence) _lseeki64(fd, offset, whence)
#define strtok_r(str, delim, save) strtok_s(str, delim, save)

typedef int socklen_t;
#define OPENREAD (O_RDONLY | O_BINARY)
#define OPENWRITE (O_WRONLY | O_BINARY)

typedef HANDLE mux_t;
#define mux_create(mux) (((mux=CreateMutex(NULL,FALSE,NULL))!=NULL)?0:-1)
#define mux_destroy(mux) ((CloseHandle(mux)!=0)?0:-1)
#define mux_lock(mux) ((WaitForSingleObject(mux,INFINITE)!=WAIT_FAILED)?0:-1)
#define mux_unlock(mux) ((ReleaseMutex(mux)!=0)?0:-1)

typedef uintptr_t thread_t;
#define start_thread(id,func,arg) (((id=_beginthread(func,0,arg))!=-1)?0:-1)
#define end_thread() _endthread()
#define thread_id() GetCurrentThreadId()
#define join_thread(id) WaitForSingleObject((HANDLE)(id), INFINITE)
#define destroy_thread(id) CloseHandle((HANDLE)(id))
#define THREAD_FUNC void
#define THREAD_RETURN return

#else  // if WINDOWS

#include <inttypes.h>
#include <pthread.h>


#define closesocket(s) close(s)

#ifdef VMS
pid_t GENERIC_SETSID(void);
#define setsid GENERIC_SETSID
#define fork vfork
typedef unsigned int socklen_t;
#define isfullpath(str) (1)
#define PATH_SEP ':'
#define open(name, flag, mode) open(name, flag, mode, "ctx=stm")
#else
#define isfullpath(str) (str[0] == '/')
#define PATH_SEP '/'
#endif

#define F_i64 "%" PRId64
#define F_u64 "%" PRIu64

typedef uint64_t f_offset_t;
typedef struct stat stat_struct;
typedef struct utimbuf utim_buf;
#define stat_func(name, buf) stat(name, buf)
#define lstat_func(name, buf) lstat(name, buf)
#define lseek_func(fd, offset, whence) lseek(fd, offset, whence)

typedef int SOCKET;
#define INVALID_SOCKET -1
#define SOCKET_ERROR -1
#ifndef INADDR_NONE
#define INADDR_NONE -1
#endif
#define OPENREAD (O_RDONLY)
#define OPENWRITE (O_WRONLY)

typedef pthread_mutex_t mux_t;
#define mux_create(mux) pthread_mutex_init(&(mux),NULL)
#define mux_destroy(mux) pthread_mutex_destroy(&(mux))
#define mux_lock(mux) pthread_mutex_lock(&(mux))
#define mux_unlock(mux) pthread_mutex_unlock(&(mux))

typedef pthread_t thread_t;
#define start_thread(id,func,arg) pthread_create(&(id),NULL,func,arg)
#define end_thread() pthread_exit(NULL)
#define thread_id() pthread_self()
#define join_thread(id) pthread_join(id, NULL)
#define destroy_thread(id)
#define THREAD_FUNC void *
#define THREAD_RETURN return NULL

#endif // if WINDOWS

#define VERSIONSTR "UFTP version 5.0.1  Copyright (C) 2001-2022  Dennis A. Bush"
#define UFTP_VER_NUM 0x50
#define UFTP4_VER_NUM 0x40

#define ANNOUNCE      1
#define REGISTER      2
#define CLIENT_KEY    3
#define REG_CONF      4
#define KEYINFO       5
#define KEYINFO_ACK   6
#define FILEINFO      7
#define FILEINFO_ACK  8
#define FILESEG       9
#define DONE         10
#define STATUS       11
#define COMPLETE     12
#define DONE_CONF    13
#define HB_REQ       14
#define HB_RESP      15
// obsolete, keeping for v4 compatibility
#define KEY_REQ      16
#define PROXY_KEY    17
#define ENCRYPTED    18
#define ABORT        19 
#define CONG_CTRL    20
#define CC_ACK       21

#define FTYPE_REG       0
#define FTYPE_DIR       1
#define FTYPE_LINK      2
#define FTYPE_DELETE    3
#define FTYPE_FREESPACE 4

#define KEY_NONE        0
// reserved, obsolete: 1-4, keeping 3-4 for v4 compatibility
#define KEY_AES128_CBC  3
#define KEY_AES256_CBC  4
#define KEY_AES128_GCM  5
#define KEY_AES256_GCM  6
#define KEY_AES128_CCM  7
#define KEY_AES256_CCM  8

#define HASH_NONE   0
// reserved, obsolete: 1-2, keeping 2 for v4 ECDH compatibility
#define HASH_SHA1   2
#define HASH_SHA256 3
#define HASH_SHA384 4
#define HASH_SHA512 5

// obsolete, keeping for v4 compatibility
#define SIG_NONE    0
#define SIG_HMAC    1
#define SIG_KEYEX   2
#define SIG_AUTHENC 3

// obsolete, keeping for v4 compatibility
#define KEYEX_NONE          0
#define KEYEX_RSA           1
#define KEYEX_ECDH_RSA      2
#define KEYEX_ECDH_ECDSA    3

#define KEYBLOB_RSA     1
#define KEYBLOB_EC      2

// reserved, obsolete: 1-22
#define CURVE_secp256r1     23
#define CURVE_secp384r1     24
#define CURVE_secp521r1     25
#define CURVE_prime256v1    CURVE_secp256r1

#define CC_NONE     0
// reserved, obsolete: 1
#define CC_TFMCC    2
#define CC_PGMCC    3

#define EXT_ENC_INFO        1
#define EXT_TFMCC_DATA_INFO 2
#define EXT_TFMCC_ACK_INFO  3

#define EXT_PGMCC_DATA_INFO 4
#define EXT_PGMCC_NAK_INFO  5
#define EXT_PGMCC_ACK_INFO  6

#define EXT_FREESPACE_INFO 7

#define FLAG_SYNC_MODE      0x01
#define FLAG_SYNC_PREVIEW   0x02
#define FLAG_IPV6           0x04

#define FLAG_CLIENT_AUTH    0x01

#define FLAG_PARTIAL        0x01

#define FLAG_CURRENT_FILE   0x01

#define FLAG_CC_CLR         0x01
#define FLAG_CC_RTT         0x02
#define FLAG_CC_START       0x04
#define FLAG_CC_LEAVE       0x08

#define COMP_STAT_NORMAL    0
#define COMP_STAT_SKIPPED   1
#define COMP_STAT_OVERWRITE 2
#define COMP_STAT_REJECTED  3

#define HB_AUTH_FAILED      0
#define HB_AUTH_OK          1
#define HB_AUTH_CHALLENGE   2

#define MAXFILENAME 100
#define MAXDIRNAME 200
#define MAXPATHNAME 300
#define MAXBACKUPPATHNAME 600
#define MAXPROXYDEST 1000
#define MAXDIR 10
#define MAXSECTION 65536

#define DESTNAME_LEN 80
#define IFNAME_LEN 25
#define PORTNAME_LEN 20
#define MAX_INTERFACES 200
#define MAXMTU 9000

#define PUBKEY_LEN 392  // big enough for a keyblob with RSA-3072
#define RAND_LEN 32     // RFC 8446/5246
#define HMAC_LEN 64     // big enough for SHA-512
#define HASH_LEN 64     // big enough for SHA-512
#define VERIFY4_LEN 12  // RFC 5246
#define MASTER4_LEN 48  // RFC 5246
#define MASTER_LEN 32   // based on SHA-256 length
#define ENC_MASTER_LEN 48 // based on SHA-256 length + tag length
#define MAXIV 16        // big enough for AES256
#define MAXKEY 32       // big enough for AES256
#define KEYBLSIZE 16    // Maximum symmetric key blocksize
#define DEF_RSA_LEN 1024 // Default length of generated RSA keys
#define RSA_MIN 1024    // Minimum RSA key length
#define RSA_MAX 3072    // Maximum RSA key length
#define DEF_CURVE CURVE_prime256v1  // Default EC curve
#define RSA_EXP 65537   // Public key exponent of generated RSA keys
#define SALT_LEN 4      // Length of salt for IV
#define GCM_IV_LEN 12   // Length of IV for ciphers in GCM mode
#define CCM_IV_LEN 12   // Length of IV for ciphers in CCM mode
#define GCM_TAG_LEN 16  // Length of tag for ciphers in GCM mode
#define CCM_TAG_LEN 16  // Length of tag for ciphers in CCM mode

#define ERR_NONE        0   // Normal exit code
#define ERR_PARAM       1   // One or more command line parameters invalid
#define ERR_SOCKET      2   // Error initializing sockets
#define ERR_CRYPTO      3   // Error initializing cryptographic keys
#define ERR_LOGGING     4   // Error while opening log file
#define ERR_ALLOC       5   // Memory allocation failure
#define ERR_INTERRUPTED 6   // Interrupted by the user or an unexpected signal
#define ERR_NO_REGISTER 7   // No client responded to ANNOUNCE with REGISTER
#define ERR_NO_FILEINFO 8   // No client responded to a FILEINFO
#define ERR_DROPPED     9   // All clients were lost or aborted
#define ERR_NO_FILES    10  // No files were sent

struct uftp_h {
    uint8_t version;
    uint8_t func;
    uint16_t seq;
    uint32_t src_id;        // ID of sender
    uint32_t group_id;
    uint8_t group_inst;     // Group restart number
    uint8_t grtt;           // RFC 5401, Unused in upstream messages
    uint8_t gsize;          // Unused in upstream messages
    uint8_t reserved;
};  // sizeof = 16

struct encrypted_h {
    uint32_t iv_ctr_hi;
    uint32_t iv_ctr_lo;
    // v4 only, set to 0 in v5
    uint16_t sig_len;
    uint16_t payload_len;
};  // sizeof = 12 + sig_len + payload_len

struct announce_h {
    uint8_t func;  // always ANNOUNCE
    uint8_t hlen;
    uint8_t flags;
    uint8_t robust;
    uint8_t cc_type;
    uint8_t reserved;
    uint16_t blocksize;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
    //uint32_t publicmcast;   // for IPv4
    //uint8_t publicmcast[16];  // for IPv6
    //uint32_t privatemcast;   // for IPv4
    //uint8_t privatemcast[16];  // for IPv6
};  // sizeof = 16 + 2*iplen + (encrypted?sizeof(enc_info_he):0) + {uint32_t[]}

struct enc_info_he {
    uint8_t exttype;   // always EXT_ENC_INFO
    uint8_t extlen;
    uint8_t flags;
    // v4 only, set to 0 in v5
    uint8_t keyextype_sigtype;  // & 0xF0 = keyextype, & 0x0F = sigtype
    uint8_t keytype;
    uint8_t hashtype;
    uint16_t keylen;
    uint16_t dhlen;
    uint16_t siglen;
    uint8_t rand1[RAND_LEN];
    //uint8_t keyblob[];
    //uint8_t dhkey[];
    //uint8_t sig[];
};  //sizeof = 44 + keylen + dhlen + siglen

struct rsa_blob_t {
    uint8_t blobtype;  // always KEYBLOB_RSA
    uint8_t reserved;
    uint16_t modlen;
    uint32_t exponent;
    //uint8_t modulus[];
};  //sizeof = 8 + modlen

struct ec_blob_t {
    uint8_t blobtype;  // always KEYBLOB_EC
    uint8_t curve;
    uint16_t keylen;
    //uint8_t key[];
};  //sizeof = 4 + keylen

struct client_key_h {
    uint8_t func;  // always CLIENT_KEY
    uint8_t hlen;
    uint16_t reserved;
    uint16_t bloblen;
    uint16_t siglen;
    //uint8_t keyblob[];
    //uint8_t verify[];
};  // sizeof = 8 + bloblen + siglen

struct register_h {
    uint8_t func;  // always REGISTER
    uint8_t hlen;
    uint16_t dhlen;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
    uint8_t rand2[RAND_LEN];
    //uint8_t dhkey[];
};  // sizeof = 44 + dhlen + {uint32_t[]}

struct regconf_h {
    uint8_t func;  // always REG_CONF
    uint8_t hlen;
    uint16_t reserved;
};  // sizeof = 4 + {uint32_t[]}

struct keyinfo_h {
    uint8_t func;  // always KEYINFO
    uint8_t hlen;
    uint16_t siglen;
    uint32_t iv_ctr_hi;
    uint32_t iv_ctr_lo;
};  // sizeof = 12 + siglen + {destkey[]}

struct destkey {
    uint32_t dest_id;
    uint8_t groupmaster[ENC_MASTER_LEN];
};  // sizeof = 52

struct keyinfoack_h {
    uint8_t func;  // always KEYINFO_ACK
    uint8_t hlen;
    uint16_t verify_len;
    //uint8_t verify_data[];
};  // sizeof = 4 + verify_len

struct fileinfo_h {
    uint8_t func;  // always FILEINFO
    uint8_t hlen;
    uint16_t file_id;
    uint8_t ftype;
    uint8_t reserved1;
    uint16_t ftstamp_hi;
    uint8_t namelen;
    uint8_t linklen;
    uint16_t hifsize;
    uint32_t lofsize;
    uint32_t ftstamp_lo;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
    //char name[MAXPATHNAME];
};  // sizeof = 28 + namelen + {uint32_t[]}

struct fileinfoack_h {
    uint8_t func;  // always FILEINFO_ACK
    uint8_t hlen;
    uint16_t file_id;
    uint8_t flags;
    uint8_t reserved1;
    uint16_t reserved2;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
};  // sizeof = 16 + {uint32_t[]}

struct fileseg_h {
    uint8_t func;  // always FILESEG
    uint8_t hlen;
    uint16_t file_id;
    uint16_t section;
    uint16_t sec_block;
};  // sizeof = 8 

// Append to FILESEG
struct tfmcc_data_info_he {
    uint8_t exttype;   // always EXT_TFMCC_DATA_INFO
    uint8_t extlen;
    uint16_t send_rate;
    uint16_t cc_seq;
    uint16_t cc_rate;
};  // sizeof = 8

// Append to FILESEG
struct pgmcc_data_info_he {
    uint8_t exttype;   // always EXT_PGMCC_DATA_INFO
    uint8_t extlen;
    uint16_t reserved;
    uint32_t acker;
};  // sizeof = 8

struct done_h {
    uint8_t func;  // always DONE
    uint8_t hlen;
    uint16_t file_id;
    uint16_t section;
    uint16_t reserved;
};  // sizeof = 8 + {uint32_t[]}

struct status_h {
    uint8_t func;  // always STATUS
    uint8_t hlen;
    uint16_t file_id;
    uint16_t section;
    uint16_t reserved;
};  // sizeof = 8 

// Append to STATUS or CC_ACK
struct tfmcc_ack_info_he {
    uint8_t exttype;   // always EXT_TFMCC_ACK_INFO
    uint8_t extlen;
    uint8_t flags;  // RTT, SS, LEAVE
    uint8_t reserved;
    uint16_t cc_seq;
    uint16_t cc_rate;
    uint32_t client_id;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
};  // sizeof = 20 

// Append to STATUS
struct pgmcc_nak_info_he {
    uint8_t exttype;   // always EXT_PGMCC_NAK_INFO
    uint8_t extlen;
    uint16_t loss;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
};  // sizeof = 12

// Append to CC_ACK
struct pgmcc_ack_info_he {
    uint8_t exttype;   // always EXT_PGMCC_ACK_INFO
    uint8_t extlen;
    uint16_t reserved;
    uint16_t loss;
    uint16_t max_seq;
    uint32_t ack_bitmap;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
};  // sizeof = 20

struct complete_h {
    uint8_t func;  // always COMPLETE
    uint8_t hlen;
    uint16_t file_id;
    uint8_t status;
    uint8_t reserved1;
    uint16_t reserved2;
};  // sizeof = 8 + {uint32_t[]}

struct freespace_info_he {
    uint8_t exttype;  // always EXT_FREESPACE_INFO
    uint8_t extlen;
    uint16_t reserved;
    uint32_t freespace_hi;
    uint32_t freespace_lo;
};  // sizeof = 12

struct doneconf_h {
    uint8_t func;  // always DONE_CONF
    uint8_t hlen;
    uint16_t reserved;
};  // sizeof = 4 + {uint32_t[]}

struct abort_h {
    uint8_t func;  // always ABORT
    uint8_t hlen;
    uint8_t flags;
    uint8_t reserved;
    uint32_t host;
    char message[300];
};  // sizeof = 308

struct hb_req_h {
    uint8_t func;  // always HB_REQ
    uint8_t hlen;
    uint16_t reserved;
    uint16_t bloblen;
    uint16_t siglen;
    uint32_t nonce;
    //uint8_t keyblob[];
    //uint8_t verify[];
};  // sizeof = 12 + bloblen + siglen

struct hb_resp_h {
    uint8_t func;  // always HB_RESP
    uint8_t hlen;
    uint8_t authenticated;
    uint8_t reserved;
    uint32_t nonce;
};  // sizeof = 8

struct key_req_h {
    uint8_t func;  // always KEY_REQ
    uint8_t hlen;
    uint16_t reserved;
};  // sizeof = 4

struct proxy_key_4_h {
    uint8_t func;  // always PROXY_KEY
    uint8_t hlen;
    uint16_t bloblen;
    uint16_t dhlen;
    uint16_t siglen;
    uint32_t nonce;
    //uint8_t keyblob[];
    //uint8_t dhkey[];
    //uint8_t verify[];
};  // sizeof = 12 + bloblen + dhlen + siglen

struct proxy_key_h {
    uint8_t func;  // always PROXY_KEY
    uint8_t hlen;
    uint16_t bloblen;
    uint16_t dhlen;
    uint16_t siglen;
    //uint8_t keyblob[];
    //uint8_t dhkey[];
    //uint8_t signature[];
};  // sizeof = 8 + bloblen + dhlen + siglen

struct cong_ctrl_h {
    uint8_t func;  // always CONG_CTRL
    uint8_t hlen;
    uint16_t reserved;
    uint16_t cc_seq;
    uint16_t cc_rate;
    uint32_t tstamp_hi;
    uint32_t tstamp_lo;
}; // sizeof = 16 + {cc_item[]}

struct cc_item {
    uint32_t dest_id;
    uint8_t flags;  // CLR, RTT, START, LEAVE
    uint8_t rtt;    // RFC 5401
    uint16_t reserved;
}; // sizeof = 8

struct cc_ack_h {
    uint8_t func;  // always CC_ACK
    uint8_t hlen;
    uint16_t reserved;
}; // sizeof = 4 + tfmcc_info_he or pgmcc_info_ack_he

#endif  // _UFTP_H
